haskell/evaluare-lenesa/schelet/lazy.hs

```diff
@@ -0,0 +1,249 @@
+{-# LANGUAGE NoMonomorphismRestriction #-}
+import Data.List
+import Debug.Trace
+import TestPP
+import Data.Char
+
+{-
+Expresia `undefined` are orice tip dar nu poate fi evaluată.
+-}
+
+{-
+    1. Să se găsească cuvintele care au lungimea cel puțin egală cu 10 caractere în două moduri:
+        1) folosind filter
+        2) folosind list comprehensions
+-}
+
+findStringsLongerThanTenChars :: [String] -> [String]
+findStringsLongerThanTenChars l = undefined
+
+findStringsLongerThanTenChars2 :: [String] -> [String]
+findStringsLongerThanTenChars2 l = undefined
+
+-- Verificare: check1
+check1 :: TestData
+check1 = tests_ 1
+    [ testVal "findStringsLongerThanTenChars" ["programatoare","interesanta"] $
+        findStringsLongerThanTenChars ["ana", "este", "o", "programatoare", "foarte", "interesanta"]
+    , testVal "findStringsLongerThanTenChars2" ["programatoare","interesanta"] $
+        findStringsLongerThanTenChars2 ["ana", "este", "o", "programatoare", "foarte", "interesanta"]
+    ]
+
+{-
+    2. Să se construiască o listă de perechi de tip (string, lungime_string) în două moduri:
+        1) folosind map
+        2) folosind list comprehensions
+-}
+
+buildPairsStringLength :: [String] -> [(String, Int)]
+buildPairsStringLength l = undefined
+
+buildPairsStringLength2 :: [String] -> [(String, Int)]
+buildPairsStringLength2 l = undefined
+
+-- Verificare: check2
+check2 :: TestData
+check2 = tests_ 2
+    [ testVal "buildPairsStringLength" [("ana",3),("este",4),("o",1), ("programatoare",13),("foarte",6),("interesanta",11)] $
+        buildPairsStringLength ["ana", "este", "o", "programatoare", "foarte", "interesanta"]
+    , testVal "buildPairsStringLength2" [("ana",3),("este",4),("o",1), ("programatoare",13),("foarte",6),("interesanta",11)] $
+        buildPairsStringLength2 ["ana", "este", "o", "programatoare", "foarte", "interesanta"]
+    ]
+
+{-
+    3. Implementați, folosind obligatoriu list-comprehensions, operații pe mulțimi:
+    intersecție, diferență, produs cartezian. Utilizați ulterior funcțiile definite anterior
+    pentru a reprezenta reuniunea mulțimilor.
+-}
+
+setIntersection :: Eq a => [a] -> [a] -> [a]
+setIntersection a b = undefined
+
+setDiff :: Eq a => [a] -> [a] -> [a]
+setDiff a b = undefined
+
+cartProduct :: [a] -> [b] -> [(a, b)]
+cartProduct a b = undefined
+
+setUnion :: Eq a => [a] -> [a] -> [a]
+setUnion a b = undefined
+
+-- Verificare: check3
+check3 :: TestData
+check3 = tests_ 3
+    [ testSet "setIntersection" [1, 2, 6] $ setIntersection a b
+    , testSet "setDiff" [3, 7] $ setDiff a b
+    , testVal "cartProduct" [(1, 3), (1, 4), (1, 5), (2, 3), (2, 4), (2, 5)] $ cartProduct [1, 2] [3, 4, 5]
+    , testSet "setUnion" [1, 2, 3, 4, 6, 7, 8, 10] $ setUnion a b
+    ]
+  where 
+      a = [1, 7, 3, 6, 2]
+      b = [2, 8, 6, 10, 4, 1]
+
+
+{-
+    9. Experimentați în consolă funcționalitățile funcției "trace", aplicată în funcțiile "evens", "squaredEvens" și "sumOfSquaredEvens".
```

Comment:
Poate ar trebui explicat și în comentarii ce face trace, și anume faptul că:
- ia 2 parametri, un mesaj și o altă expresie
- rezultatul este dat de al doilea parametru
- mesajul e afișat în momentul în care se forțează evaluarea aplicației lui trace